### PR TITLE
fix: revert to original nodes if `evaluate` is failed

### LINKED ISSUE #15: Do not bail-out when a static problem is found
A file can have a mix of dynamic and static `fs.readFileSync`, and when a dynamic read is found, throws and error and stops the build.

Simply, if a dynamic read is found, it should continue with the next one, as is not a terminal error actually.

I've modified `static-eval.js` it to achieve this with:
image

Is a ugly workaround, but I had to test if this works. It worked.

Maybe there's a better way to achieve this.

### PR DESCRIPTION
**Pull request by azu**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Code style update
- [ ] Refactor (refactoring or adding test which isn't a fix or add a feature)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No (Probably)

**Did you test your solution?**

- [ ] I lightly tested it in one browser
- [ ] I deeply tested it in several browsers
- [x] I wrote tests around it (unit tests, integration tests, E2E tests)

## Problem Description

fix #15 

## Solution Description

1. Pre-save original `path` and `node` before inlining
2. When `evaluate()` throw error, revert to original `node`

## Side Effects, Risks, Impact

- [ ] N/A

**Aditional comments:**

```js
const fs = require('fs');
const path = require('path');
const str = fs.readFileSync(path.join(__dirname, '/hello.txt'), 'utf8');
console.log(str);

function evaluated (blah) {
  const result = fs.readFileSync(path.join(__dirname, blah), 'utf8');
  console.log(result);
}

setTimeout(() => evaluated('/hello.txt'), 1500);
const str2 = fs.readFileSync(path.join(__dirname, '/hello.txt'), 'utf8');
console.log(str2);
```

to be

```js
const fs = require('fs');

const path = require('path');

const str = "hello";
console.log(str);

function evaluated(blah) {
  const result = fs.readFileSync(path.join(__dirname, blah), 'utf8');
  console.log(result);
}

setTimeout(() => evaluated('/hello.txt'), 1500);
const str2 = "hello";
console.log(str2);
```

### CODE REVIEW

File: lib/static-eval.js [hunk 1]
```diff
@@ -23,90 +23,107 @@ module.exports = evaluate;
...
 
-  // Replace identifiers with known values
-  path.traverse({
-    Identifier: function (ident) {
-      var key = ident.node.name;
-      if (key in vars) {
-        ident.replaceWith(t.valueToNode(vars[key]));
+  // save replaced nodes for reverting nodes
+  const replacedMap = new Map();
+  const replaceWith = (path, node) => {
```
Comment (azu) [inline lib/static-eval.js:28]: `path.replaceWith` wrapper function for save original node

File: lib/static-eval.js [hunk 2]
```diff
@@ -23,90 +23,107 @@ module.exports = evaluate;
...
+    replacedMap.set(path, {
+      oldNode: path.node,
+      newNode: node,
+    });
+    path.replaceWith(node);
+  };
+  try {
```
Comment (azu) [inline lib/static-eval.js:36]: wrap with `try...catch`

File: lib/static-eval.js [hunk 3]
```diff
@@ -23,90 +23,107 @@ module.exports = evaluate;
...
+      if (!result.confident) {
+        throw new Error('Not able to statically evaluate the expression(s) for babel-plugin-static-fs.\n' +
+            'Try changing your source code to something that can be evaluated at build-time, e.g.\n' +
+            "    const src = fs.readFileSync(__dirname + '/foo.txt', 'utf8');\n");
+      }
+      return result.value;
+    }
+  } catch (error) {
+    // revert to original nodes if evaluating error is happened
+    for (const [path, nodes] of replacedMap.entries()) {
```
Comment (azu) [inline lib/static-eval.js]: If `evaluate` throw error, revert to original node.

File: lib/static-eval.js [hunk 4]
```diff
@@ -23,90 +23,107 @@ module.exports = evaluate;
...
+      path.replaceWith(nodes.oldNode);
     }
-    return result.value;
+    return error;
```
Comment (azu) [inline lib/static-eval.js:128]: return error to prevent to delete `const fs = require("fs")`

File: index.js [hunk 5]
```diff
@@ -126,12 +126,10 @@ module.exports = function (babel) {
...
           // e.g. fs.readFileSync(...) -> 'foobar'
-          // try {
-            evaluate(state.opts, path, state.file.opts.filename);
-          // } catch (err) {
-          //   console.error(err);
-          //   errors.push(err);
-          // }
+          const result = evaluate(state.opts, path, state.file.opts.filename);
+          if (result instanceof Error) {
+            errors.push(result);
```
Comment (azu) [inline index.js:131]: If `evaluate` return an Error, do not delete `require("fs")`

**Review by azu — COMMENTED** [review]
@mattdesl I've implemented restore mechanism in `evalute` function.
Do you think about it?


"No whitespace" view help to watch
`https://github.com/Jam3/babel-plugin-static-fs/pull/18?w=1`

**Comment by mattdesl** [comment]
Thanks! Seems great. Is there any edge case where evaluate would actually statically eval to an error? This is why I was using try/catch to detect whether the eval succeeded, rather than your approach of checking the result of the evaluate.

**Comment by azu** [comment]
>  Is there any edge case where evaluate would actually statically eval to an error?

https://github.com/babel/babel/blob/e541d6031af4008cb649d3408f7387fbb4106bc8/packages/babel-traverse/src/path/evaluation.js
https://github.com/babel/babel/blob/master/packages/babel-traverse/test/evaluation.js

I think that `path.evalute()` does not throw error in nomal scenario.
But, `try { ...inlining... } catch (error) { ...restore... }` is simple.

**Comment by mattdesl** [comment]
Ok, I ended up pulling this in and published as 3.0.0 since its a breaking change (gracefully skipping dynamic rather than bailing out early). Thanks again.

I've changed it slightly to use try/catch instead of checking the return type as I feel the code is clearer (error remains an 'exceptional case' rather than just another return value). I've also added a `{ dynamic }` flag (default true) which can be set to false if users wish to disable this feature and go back to bailing out on dynamic expressions.

Cheers!

**Pull request merged**